src/System.Linq/src/System/Linq/Take.cs

```diff
@@ -96,5 +97,54 @@ private static IEnumerable<TSource> TakeWhileIterator<TSource>(IEnumerable<TSour
                 yield return element;
             }
         }
+
+        public static IEnumerable<TSource> TakeLast<TSource>(this IEnumerable<TSource> source, int count)
+        {
+            if (source == null)
+            {
+                throw Error.ArgumentNull(nameof(source));
+            }
+
+            if (count <= 0)
+            {
+                return EmptyPartition<TSource>.Instance;
+            }
+
+            return TakeLastIterator(source, count);
+        }
+
+        private static IEnumerable<TSource> TakeLastIterator<TSource>(IEnumerable<TSource> source, int count)
+        {
+            Debug.Assert(source != null);
+            Debug.Assert(count > 0);
+
+            var queue = new Queue<TSource>();
+
+            using (IEnumerator<TSource> e = source.GetEnumerator())
+            {
+                while (e.MoveNext())
+                {
+                    if (queue.Count < count)
+                    {
+                        queue.Enqueue(e.Current);
+                    }
+                    else
+                    {
+                        do
+                        {
+                            queue.Dequeue();
+                            queue.Enqueue(e.Current);
+                        }
+                        while (e.MoveNext());
+                        break;
+                    }
+                }
+            }
+
+            while (queue.Count > 0)
+            {
+                yield return queue.Dequeue();
+            }
```

Comment:
If this was `foreach(TSource item in queue) yield return queue;` then the lack of updating of `queue` on dequeuing might pay off.

Perhaps even:

```C#
private class TakeLastIterator<TSource> : IEnumerable<TSource>
{
	private readonly IEnumerable<TSource> _source;
	private readonly int _count;
	
	public TakeLastIterator(IEnumerable<TSource> source, int count)
	{
		_source = source;
		_count = count;
	}
	
	public IEnumerator<TSource> GetEnumerator()
	{
	    var queue = new Queue<TSource>();
		int count = _count;
	
	    using (IEnumerator<TSource> e = _source.GetEnumerator())
	    {
	        while (e.MoveNext())
	        {
	            if (count-- != 0)
	            {
	                queue.Enqueue(e.Current);
	            }
	            else
	            {
	                do
	                {
	                    queue.Dequeue();
	                    queue.Enqueue(e.Current);
	                }
	                while (e.MoveNext());
	                break;
	            }
	        }
		}
		
		return queue.GetEnumerator();
	}
	
	IEnumerator IEnumerable.GetEnumerator()
	{
		return GetEnumerator();
	}
}
```

And then a level of indirection is removed entirely, but the implementing `queue` is still not exposed.